lasagne/layers/recurrent.py

```diff
@@ -611,42 +602,59 @@ def step(input_dot_W_n, cell_previous, hid_previous):
 
-            cell, hid = step(input_dot_W_n, cell_previous, hid_previous)
+            cell, hid = step(input_dot_w_n, cell_previous, hid_previous)
 
             # If mask is 0, use previous state until mask = 1 is found.
             # This propagates the layer initial state when moving backwards
             # until the end of the sequence is found.
-            not_mask = 1 - mask
-            cell = cell*mask + cell_previous*not_mask
-            hid = hid*mask + hid_previous*not_mask
+            not_mask = 1 - mask_n
+            cell = cell*mask_n + cell_previous*not_mask
+            hid = hid*mask_n + hid_previous*not_mask
 
             return [cell, hid]
 
-        if self.backwards and mask is not None:
+        if mask is not None:
             # mask is given as (batch_size, seq_len). Because scan iterates
             # over first dimension, we dimshuffle to (seq_len, batch_size) and
             # add a broadcastable dimension
             mask = mask.dimshuffle(1, 0, 'x')
-            sequences = [input_dot_W, mask]
+            sequences = [input_dot_w, mask]
             step_fun = step_masked
         else:
-            sequences = input_dot_W
+            sequences = input_dot_w
             step_fun = step
 
         # Scan op iterates over first dimension of input and repeatedly
         # applies the step function
-        output = theano.scan(step_fun, sequences=sequences,
-                             outputs_info=[self.cell_init, self.hid_init],
-                             go_backwards=self.backwards,
-                             truncate_gradient=self.gradient_steps)[0][1]
-
-        # Now, dimshuffle back to (n_batch, n_time_steps, n_features))
-        output = output.dimshuffle(1, 0, 2)
-
-        # if scan is backward reverse the output
-        if self.backwards:
-            output = output[:, ::-1, :]
+        cell_out, hid_out = theano.scan(
+            step_fun,
+            sequences=sequences,
+            outputs_info=[self.cell_init, self.hid_init],
+            go_backwards=self.backwards,
+            truncate_gradient=self.gradient_steps)[0]
+
+        if self.return_sequence:
+            # dimshuffle back to (n_batch, n_time_steps, n_features))
+            hid_out = hid_out.dimshuffle(1, 0, 2)
+
+            # if scan is backward reverse the output
+            if self.backwards:
+                hid_out = hid_out[:, ::-1, :]
+        else:
+            # no need to dimshuffle because we return the last state
+            hid_out = hid_out[-1]
+        output = hid_out
+
+        # repeat for cell values if needed
+        if self.return_cell:
+            if self.return_sequence:
+                cell_out = cell_out.dimshuffle(1, 0, 2)
+                if self.backwards:
+                    cell_out = cell_out[:, ::-1, :]
+            else:
+                cell_out = cell_out[-1]
+            output = [cell_out, output]
```

Comment:
At the moment I don't think there is anything in Lasagne that relies on a layer's output being a Theano expression - except of course other layers. So that would mean it's not necessarily a problem to return other data types from `get_output_for()`.

However, you would then need a special set of layers that can take lists of expressions as input. (it would not be possible to stack a dense layer on top of this for example - it doesn't make sense). That sounds like it could get complicated. Maybe there is a better solution.

I don't fully understand why this is necessary to be able to build an encoder/decoder model, though. Why do you need the cell values for that? I don't really understand the alternatives you proposed either, but that's probably down to not knowing enough about LSTMs. I should go and read up first and then come back :)